### ISSUE TITLE
Make filter() use TypeGuard

### ISSUE THREAD

**Issue by JelleZijlstra** [python/typeshed#5661]
This will probably require a lot of overloads.

**Comment by layday** [comment]
Does it not require just the one extra overload?

```py
@overload
def __init__(self, __function: Callable[[_S], TypeGuard[_T]], __iterable: Iterable[_S]) -> None: ...
```

**Comment by AlexWaygood** [comment]
Fixed by https://github.com/python/typeshed/issues/5661

**Comment by laundmo** [comment]
> Fixed by #5661

This links to the current issue, not the PR implementing this.

**Comment by AlexWaygood** [comment]
> > Fixed by #5661
> 
> This links to the current issue, not the PR implementing this.

Thanks, I meant #6726.
